gcc/rust/checks/errors/rust-feature.h

```diff
@@ -0,0 +1,73 @@
+// Copyright (C) 2020-2022 Free Software Foundation, Inc.
+
+// This file is part of GCC.
+
+// GCC is free software; you can redistribute it and/or modify it under
+// the terms of the GNU General Public License as published by the Free
+// Software Foundation; either version 3, or (at your option) any later
+// version.
+
+// GCC is distributed in the hope that it will be useful, but WITHOUT ANY
+// WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+// for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with GCC; see the file COPYING3.  If not see
+// <http://www.gnu.org/licenses/>.
+
+#ifndef RUST_FEATURE_H
+#define RUST_FEATURE_H
+
+#include "rust-session-manager.h"
+#include "rust-optional.h"
+
+namespace Rust {
+
+class Feature
+{
+public:
+  enum class State
+  {
+    ACCEPTED,
+    ACTIVE,
+    REMOVED,
+    STABILIZED,
+  };
+
+  enum class Name
+  {
+    ASSOCIATED_TYPE_BOUNDS,
+    INTRINSICS,
+    RUSTC_ATTRS,
+    DECL_MACRO,
+  };
+
+  const std::string &as_string () { return m_name_str; }
+  Name name () { return m_name; }
+  const std::string &description () { return m_description; }
+  State state () { return m_state; }
+
+  static Optional<Name> as_name (const std::string &name);
+  static Feature create (Name name);
+
+private:
+  Feature (Name name, State S, const char *N, const char *RS, uint64_t I,
+	   const Optional<CompileOptions::Edition> &E, const char *D)
+    : m_state (S), m_name (name), m_name_str (N), m_rustc_since (RS), issue (I),
+      edition (E), m_description (D)
+  {}
```

Comment:
@TuringKi the argument names to the function are still not precise enough IMO
```suggestion
  Feature (Name name, State state, const char *name_str, const char *rustc_since, uint64_t issue_number,
	   const Optional<CompileOptions::Edition> &edition, const char *description)
    : m_state (state), m_name (name), m_name_str (name_str), m_rustc_since (rustc_since), issue (issue_number),
      edition (edition), m_description (description)
  {}
```